### PR TITLE
MM-62045 Enable eslint-plugin-jsx-a11y and fix occurrences of jsx-a11y/anchor-has-content

### PR DESCRIPTION
**Pull request by hmhealey**
#### Summary
As part of trying to reach compliance with WCAG AA, we're turning on an ESLint plugin that we've had for a while to improve our accessibility. I've turned on warnings for things that are known issues and errors for things that aren't issues in our code base, and I'll be reviewing the other errors over the next week.

This also includes fixes for the 3 `jsx-a11y/anchor-has-content` checks that were failing as well as 3 accompanying `jsx-a11y/anchor-is-valid` checks that were also failing. I'd hoped those would've been easier since there were only 3 of them, but they were indicators of deeper issues which I addressed. See the individual commits for more details on those since I also explained more about the changes in those commit messages.

#### Ticket Link
MM-62045

#### Screenshots
I had to change some CSS and HTML for this. Most of the changes shouldn't be noticeable, but I think the ones that are noticeable look better.
|Before|After|
|-|-|
|The spacing in this one looked super weird before since, for settings that aren't overridden by environment variables, there were large spaces between each of these checkboxes before.<br>Screenshot 2024-12-05 at 12 26 56 PM<br>Screenshot 2024-12-02 at 5 14 59 PM|Screenshot 2024-12-05 at 12 26 36 PM<br>Screenshot 2024-12-02 at 4 30 29 PM|
|Screenshot 2024-12-02 at 5 14 37 PM|Screenshot 2024-12-02 at 5 13 43 PM|
Screenshot 2024-12-02 at 5 15 10 PM|Screenshot 2024-12-02 at 5 13 15 PM

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Review by lieut-data — APPROVED** [review]
Awesome to see the linter getting engaged!

File: webapp/channels/src/components/admin_console/admin_definition.tsx [hunk 1]
```diff
@@ -761,7 +761,7 @@ const AdminDefinition: AdminDefinitionType = {
...
-                            disabled_help_text: defineMessage({id: 'admin.service.useLetsEncryptDescription.disabled', defaultMessage: "Enable the automatic retrieval of certificates from Let's Encrypt. The certificate will be retrieved when a client attempts to connect from a new domain. This will work with multiple domains. This setting cannot be enabled unless the [Forward port 80 to 443](#SystemSettings.Forward80To443) setting is set to true."}),
+                            disabled_help_text: defineMessage({id: 'admin.service.useLetsEncryptDescription.disabled', defaultMessage: "Enable the automatic retrieval of certificates from Let's Encrypt. The certificate will be retrieved when a client attempts to connect from a new domain. This will work with multiple domains. This setting cannot be enabled unless the [Forward port 80 to 443](#ServiceSettings.Forward80To443) setting is set to true."}),
```
Comment (larkox) [inline webapp/channels/src/components/admin_console/admin_definition.tsx:764]: Is this change intended? Either way, probably the link should not be in the text.

File: webapp/channels/src/i18n/en.json [hunk 2]
```diff
@@ -2398,7 +2398,7 @@
   "admin.service.tlsKeyFileDescription": "The private key file to use.",
   "admin.service.useLetsEncrypt": "Use Let's Encrypt:",
   "admin.service.useLetsEncryptDescription": "Enable the automatic retrieval of certificates from Let's Encrypt. The certificate will be retrieved when a client attempts to connect from a new domain. This will work with multiple domains.",
-  "admin.service.useLetsEncryptDescription.disabled": "Enable the automatic retrieval of certificates from Let's Encrypt. The certificate will be retrieved when a client attempts to connect from a new domain. This will work with multiple domains.\n \nThis setting cannot be enabled unless the [Forward port 80 to 443](#SystemSettings.Forward80To443) setting is set to true.",
+  "admin.service.useLetsEncryptDescription.disabled": "Enable the automatic retrieval of certificates from Let's Encrypt. The certificate will be retrieved when a client attempts to connect from a new domain. This will work with multiple domains.\n \nThis setting cannot be enabled unless the [Forward port 80 to 443](#ServiceSettings.Forward80To443) setting is set to true.",
```
Comment (larkox) [inline webapp/channels/src/i18n/en.json:2401]: Same comment about the potential unintended change and whether we should extract the link out of there.

**Review by larkox — APPROVED** [review]
Great work. A few nitpicks here and there, but I think nothing is blocking.

File: webapp/channels/src/components/admin_console/setting.tsx [hunk 3]
```diff
@@ -12,28 +11,22 @@ export type Props = {
     children?: React.ReactNode;
     helpText?: React.ReactNode;
     setByEnv?: boolean;
-    nested?: boolean;
 }
 
-const Settings = ({children, setByEnv, helpText, inputId, label, nested = false}: Props) => {
+const Settings = ({children, setByEnv, helpText, inputId, label}: Props) => {
```
[the remark this answers is not in the document]
Comment (hmhealey) [reply webapp/channels/src/components/admin_console/setting.tsx:16]: Yeah, they're needed for two different things. `SettingSet` is used for radio buttons or checkboxes which should be rendered in a `fieldset` and `Setting` is needed for other inputs which aren't grouped together. They take the same props, but the resulting HTML is very different

File: webapp/channels/src/components/admin_console/checkbox_setting.tsx [hunk 4]
```diff
@@ -26,29 +24,8 @@ export default class CheckboxSetting extends React.PureComponent<Props> {
     };
 
     public render() {
```
[the remark this answers is not in the document]
Comment (hmhealey) [reply webapp/channels/src/components/admin_console/checkbox_setting.tsx:26]: No, we shouldn't do that here. Despite the name of this component, it doesn't actually render a whole setting on its own. It only renders a single checkbox to be used inside of a `SettingSet`

File: webapp/channels/src/components/admin_console/setting_set.tsx [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+import React from 'react';
+
+import SetByEnv from './set_by_env';
+import type {Props as SettingProps} from './setting';
+
+export type Props = Omit<SettingProps, 'nested'>;
+
+export default function SettingSet({
+    children,
```
[the remark this answers is not in the document]
Comment (hmhealey) [reply webapp/channels/src/components/admin_console/setting_set.tsx:10]: I'm not sure I understand that here. I can understand not passing `children` when we need something more structured (particularly for plugins) or when we would use `cloneElement`, but I don't see why we'd pass something other than `children` here

File: webapp/channels/src/i18n/en.json (hunk 2 above)
Comment (hmhealey) [reply webapp/channels/src/i18n/en.json:2401]: See above

File: webapp/channels/src/components/admin_console/setting_set.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (larkox) [reply webapp/channels/src/components/admin_console/setting_set.tsx:10]: If we have children, it is up to the developer to think "what I want inside". Since these are for settings, what is inside should be quite structured: a setting. Leaving that up to the developer, makes it easier to forget / break things (at least visually).

I know we have history of adding "custom settings" all over the place, and may be not worth it to confine this. But I would love to see us using just "CheckboxSetting", "BooleanSetting", "TextSetting", etc... components, or a "Setting" component with a prop type, that will show different props depending on what type is using.

[the remark this answers is not in the document]
Comment (hmhealey) [reply webapp/channels/src/components/admin_console/setting_set.tsx:10]: Ah, I see what you mean. I think there's a difference between this and those other settings since this is a layout component for the other `XSetting` components and for any custom settings to use. It's already used by `BooleanSetting` or some `RadioSetting` as well. I'd say that `CheckboxSetting` could use it, but that continues to be a weird case since it's actually multiple config settings under the hood, so I'm not going to go through the headache of unravelling all of that right now.

I think we've actually been successful with most of those components since we do use `BooleanSetting` and stuff for the vast majority of config settings in the System Console. I'd really like to see us do the same thing in the team and user settings as well when we get a chance to

**Comment by yasserfaraazkhan** [comment]
@hmhealey can you check the e2e report ?

**Review by abhijit-singh — APPROVED** [review]
Thanks @hmhealey! What you have is already an improvement from earlier. My one suggestion would be that the space between the checkbox and the notice corresponding to it just below should be **less** than the space between the notice and the next checkbox. Just to make the grouping of which notice is for which checkbox super clear. 

 
image

**Comment by hmhealey** [comment]
> @hmhealey can you check the e2e report ?

Yeah, I'll take a look

> Thanks @hmhealey! What you have is already an improvement from earlier. My one suggestion would be that the space between the checkbox and the notice corresponding to it just below should be **less** than the space between the notice and the next checkbox. Just to make the grouping of which notice is for which checkbox super clear.

I'll take a look to see what I can do there. This case is a bit weird compared to everywhere else that the "This setting has been set through an environment variable" appears since this is the only time it appears within a list as opposed to after a group of settings

**Comment by hmhealey** [comment]
@abhijit-singh I updated the spacing (and put those settings into a `ul` if you care about the underlying HTML at all)
|Before|After|
-|-
Screenshot 2024-12-09 at 4 15 44 PM|Screenshot 2024-12-09 at 4 19 38 PM

@yasserfaraazkhan I think I've fixed all of those. My changes had a typo, and I had to update some E2E tests to refer to components differently. I didn't have time to run all the Elasticsearch tests locally, but enough worked that I trust they'll pass when we rerun the test suite. I'll do that tomorrow morning my time.

**Review by yasserfaraazkhan — APPROVED** [review]
